scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java

```diff
@@ -442,58 +443,71 @@ public final void processMessage(Record record) throws HandshakeException {
-							if (parameter == null) {
-								throw new IllegalStateException("handshake parameter are required!");
-							}
-							messageToProcess = ((GenericHandshakeMessage) messageToProcess)
-									.getSpecificHandshakeMessage(parameter);
+		int epoch = session.getReadEpoch();
+		if (epoch != record.getEpoch()) {
+			LOGGER.debug("Discarding HANDSHAKE message with wrong epoch received from peer [{}]:{}{}",
+					record.getPeerAddress(), StringUtil.lineSeparator(), record);
+			throw new IllegalArgumentException("processing record with wrong epoch! " + record.getEpoch() + " expected " + epoch);
+		}
+		try {
+			DTLSMessage messageToProcess = inboundMessageBuffer.getNextMessage(record);
+			while (messageToProcess != null) {
+				if (messageToProcess instanceof FragmentedHandshakeMessage) {
+					messageToProcess = handleFragmentation((FragmentedHandshakeMessage) messageToProcess);
+				}
+				if (messageToProcess == null) {
+					// messageToProcess is fragmented and not all parts have been received yet
+				} else {
+					// continue with the now fully re-assembled message
+					if (messageToProcess instanceof GenericHandshakeMessage) {
+						HandshakeParameter parameter = session.getParameter();
+						if (parameter == null) {
+							throw new IllegalStateException("handshake parameter are required!");
 						}
-						if (messageToProcess.getContentType() == ContentType.HANDSHAKE) {
-							// only cancel on HANDSHAKE messages
-							// the very last flight CCS + FINISH
-							// must be not canceled before the FINISH
-							DTLSFlight flight = pendingFlight.get();
-							if (flight != null) {
-								LOGGER.debug("response for flight {} started", flight.getFlightNumber());
-								flight.setResponseStarted();
-							}
+						messageToProcess = ((GenericHandshakeMessage) messageToProcess)
+								.getSpecificHandshakeMessage(parameter);
+					}
+					if (messageToProcess.getContentType() == ContentType.HANDSHAKE) {
+						// only cancel on HANDSHAKE messages
+						// the very last flight CCS + FINISH
+						// must be not canceled before the FINISH
+						DTLSFlight flight = pendingFlight.get();
+						if (flight != null) {
+							LOGGER.debug("response for flight {} started", flight.getFlightNumber());
+							flight.setResponseStarted();
 						}
-						doProcessMessage(messageToProcess);
 					}
+					doProcessMessage(messageToProcess);
+				}
 
-					// process next expected message (if available yet)
-					messageToProcess = inboundMessageBuffer.getNextMessage();
+				// process next expected message (if available yet)
+				messageToProcess = inboundMessageBuffer.getNextMessage();
+			}
+			session.markRecordAsRead(epoch, record.getSequenceNumber());
+			if (epoch < session.getReadEpoch()) {
+				final SerialExecutor serialExecutor = connection.getExecutor();
+				final List<Record> records = takeDeferredRecords();
+				for (Record deferredRecord : records) {
+					if (serialExecutor != null) {
```

Comment:
What are the benefit to create a new task ?
maybe to let a chance to handle record from other connection ?

Reply:
That's the same solution as in some other locations e.g. `receiveNextDatagramFromNetwork`.

The intention is to create smaller jobs, that the execution service doesn't block a thread for longer, and so, yes, have "chance to handle record from other connection".